solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java

```diff
@@ -211,23 +216,33 @@ public ShardResponse takeCompletedIncludingErrors() {
    */
   @Override
   public ShardResponse takeCompletedOrError() {
-    return take(true);
+    return take(true, -1);
   }
 
-  private ShardResponse take(boolean bailOnError) {
+  private ShardResponse take(boolean bailOnError, long maxAllowedTimeInMillis) {
     try {
-      while (pending.get() > 0) {
-        ShardResponse rsp = responses.take();
-        responseFutureMap.remove(rsp);
```

Comment:
Following the change, we don't remove anymore completed requests from `responseFutureMap`. My understanding is this line should be restored, otherwise there `cancel()` method will consider the completed sub-requests too.